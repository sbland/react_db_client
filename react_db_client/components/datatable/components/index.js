// export { DataTableCellHoverWrap } from './DataTableCell';
export { DataTableTotals } from './DataTableTotals';
export { DataTableHeadings } from './DataTableHeadings';
