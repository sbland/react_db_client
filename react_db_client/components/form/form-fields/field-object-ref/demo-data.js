export const defaultVal = '123';

export const demoResults = [
  {
    _id: '123',
    uid: 'foo',
    label: 'Foo',
    name: 'FooAlt',
  },
  {
    _id: '456',
    uid: 'bar',
    label: 'Bar',
    name: 'BarAlt',
  },
];
