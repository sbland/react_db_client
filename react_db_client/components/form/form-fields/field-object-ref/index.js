export { FieldObjectRef } from './field-object-ref';
