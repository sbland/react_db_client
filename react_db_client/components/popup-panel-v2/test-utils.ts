import { within, screen, waitFor } from '@testing-library/react';

export const getPopupContentByTitle = async (title) => {
  await waitFor(() => {
    screen.getAllByTestId('rdc-popupPanel');
    expect(
      screen
        .getAllByTestId('rdc-popupPanel')
        .find((p) => within(p).queryByText(title))
    ).toBeInTheDocument();
  });
  return screen
    .getAllByTestId('rdc-popupPanel')
    .find((p) => within(p).queryByText(title));
};
