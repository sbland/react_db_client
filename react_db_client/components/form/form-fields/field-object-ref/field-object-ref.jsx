import React, { useCallback, useMemo } from 'react';
import PropTypes from 'prop-types';

import { SearchAndSelectDropdown } from '@samnbuk/react_db_client.components.search-and-select-dropdown';

// const searchFn = (collection, schema, sortBy) => async (filters) => {
//   return apiGetDocuments(collection, filters || [], schema, sortBy);
// };

const parseVal = (val) => {
  if (!val) {
    return [];
  }
  return Array.isArray(val) ? val : [val];
};

const getSearchFieldPlaceholder = (val, field) => {
  return val[0] ? val[0][field] : 'Search...';
};

const SelectedItems = ({ items, labelField, handleItemClick }) => {
  return (
    <div className="">
      {items.map((item) => (
        <button
          type="button"
          className="button-one"
          onClick={() => handleItemClick(item.uid, item)}
        >
          {item[labelField]}
        </button>
      ))}
    </div>
  );
};

/**
 * A form field that deals with object id references
 */
export const FieldObjectRef = ({
  uid,
  unit,
  updateFormData,
  value,
  multiple,
  collection,
  required,
  labelField, // The field in the returned data to use as the label
  allowEmptySearch,
  searchFn,
}) => {
  let valueChecked = useMemo(() => parseVal(value), [value]);
  let searchFieldPlaceholder = multiple
    ? 'Choose from dropdown...'
    : getSearchFieldPlaceholder(valueChecked, labelField);

  const handleSelect = useCallback(
    // eslint-disable-next-line no-unused-vars
    (selectedId, data) => {
      console.log(selectedId);
      console.log(data);
      const newData = multiple
        ? [...valueChecked, data].filter(
            (v, i, self) => self.findIndex((vv) => vv.uid == v.uid) == i
          )
        : data;
      updateFormData(uid, newData);
      // if (newVal !== value) updateFormData(uid, newVal);
    },
    [updateFormData, uid, multiple, valueChecked]
  );

  const handleSelectedClick = useCallback(
    (itemUid) => {
      const newData = valueChecked.filter((v) => v.uid !== itemUid);
      updateFormData(uid, newData);
    },
    [updateFormData, uid, valueChecked]
  );

  return (
    <>
      <SearchAndSelectDropdown
        searchFunction={searchFn(collection, '_id')}
        handleSelect={handleSelect}
        selectionOverride={valueChecked}
        allowMultiple={multiple}
        returnFieldOnSelect="_id"
        searchFieldTargetField={labelField}
        labelField={labelField}
        // TODO: pass classname to sas
        className="formFieldInput"
        searchFieldPlaceholder={searchFieldPlaceholder}
        // onChange={(e) => updateFormData(uid, e.target.value)}
        required={required}
        allowEmptySearch={allowEmptySearch}
      />
      {/* TODO: Add selected items here if using multiple */}
      {multiple && (
        <SelectedItems
          items={valueChecked}
          labelField={labelField}
          handleItemClick={handleSelectedClick}
        />
      )}
      {unit && <span>{unit}</span>}
    </>
  );
};

const objRefShape = {
  _id: PropTypes.string.isRequired,
  uid: PropTypes.string.isRequired,
  name: PropTypes.string,
  label: PropTypes.string,
};

FieldObjectRef.propTypes = {
  uid: PropTypes.string.isRequired,
  unit: PropTypes.string,
  value: PropTypes.oneOfType([
    PropTypes.shape(objRefShape),
    PropTypes.arrayOf(PropTypes.shape(objRefShape)),
  ]),
  updateFormData: PropTypes.func.isRequired,
  multiple: PropTypes.bool,
  required: PropTypes.bool,
  labelField: PropTypes.string,
  collection: PropTypes.string.isRequired,
  allowEmptySearch: PropTypes.bool,
  searchFn: PropTypes.func.isRequired,
};

FieldObjectRef.defaultProps = {
  unit: '',
  value: null,
  multiple: false,
  // defaultVal: null,
  required: false,
  labelField: 'label',
  allowEmptySearch: false,
};
