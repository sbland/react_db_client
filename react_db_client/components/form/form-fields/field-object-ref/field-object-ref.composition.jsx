import React, { useState } from 'react';
import { CompositionWrapDefault } from '@react_db_client/helpers.composition-wraps';
import { demoResults } from './demo-data';
import { FieldObjectRef } from './field-object-ref';

export const BasicFieldObjectRef = () => {
  const [value, setValue] = useState(null);
  const updateFormData = (uid, newVal) => {
    console.log(uid);
    console.log(newVal);
    setValue(newVal);
  };
  return (
    <div>
      <CompositionWrapDefault height="4rem" width="4rem">
        <FieldObjectRef
          uid="demoid"
          unit="demounit"
          updateFormData={updateFormData}
          value={value}
          multiple={false}
          required={false}
          labelField="label"
          collection="democollection"
          asyncGetDocuments={async (collection, filters, schema, sortBy) => demoResults}
          allowEmptySearch
        />
      </CompositionWrapDefault>
      Value: {value && value.label}
    </div>
  );
};

export const Multiple = () => {
  const [value, setValue] = useState(null);
  const updateFormData = (uid, newVal) => {
    console.log(uid);
    console.log(newVal);
    setValue(newVal);
  };
  return (
    <div>
      <CompositionWrapDefault height="4rem" width="4rem">
        <FieldObjectRef
          uid="demoid"
          unit="demounit"
          updateFormData={updateFormData}
          value={value}
          multiple
          required={false}
          labelField="label"
          asyncGetDocuments={async (collection, filters, schema, sortBy) => demoResults}
          allowEmptySearch
          collection="democollection"
        />
      </CompositionWrapDefault>
      {value && value.map((v) => v.label)}
    </div>
  );
};

export const MultipleAltLabel = () => {
  const [value, setValue] = useState(null);
  return (
    <CompositionWrapDefault height="4rem" width="4rem">
      <FieldObjectRef
        uid="demoid"
        unit="demounit"
        updateFormData={alert}
        value={value}
        multiple
        required={false}
        labelField="name"
        asyncGetDocuments={async (collection, filters, schema, sortBy) => demoResults}
        allowEmptySearch
        collection="democollection"
      />
    </CompositionWrapDefault>
  );
};
