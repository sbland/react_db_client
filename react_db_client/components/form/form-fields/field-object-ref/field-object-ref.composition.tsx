import React, { useState } from 'react';
import { CompositionWrapDefault } from '@samnbuk/react_db_client.helpers.composition-wraps';

import { FieldObjectRef } from './field-object-ref';

const demoResults = [
  {
    uid: 'foo',
    label: 'Foo',
    name: 'FooAlt',
  },
  {
    uid: 'bar',
    label: 'Bar',
    name: 'BarAlt',
  },
];

export const BasicFieldObjectRef = () => {
  const [value, setValue] = useState(null);
  const updateFormData = (uid, newVal) => {
    console.log(uid);
    console.log(newVal);
    setValue(newVal);
  }
  return (
    <div>
      <CompositionWrapDefault height="4rem" width="4rem">
        <FieldObjectRef
          uid="demoid"
          unit="demounit"
          updateFormData={updateFormData}
          value={value}
          multiple={false}
          required={false}
          labelField="label"
          collection="democollection"
          searchFn={(collection, schema, sortBy) => async (filters) => demoResults}
          allowEmptySearch
        />
      </CompositionWrapDefault>
      Value: {value && value.label}
    </div>
  );
};



export const Multiple = () => {
  const [value, setValue] = useState(null);
  const updateFormData = (uid, newVal) => {
    console.log(uid);
    console.log(newVal);
    setValue(newVal);
  }
  return (
    <div>
      <CompositionWrapDefault height="4rem" width="4rem">
        <FieldObjectRef
          uid="demoid"
          unit="demounit"
          updateFormData={updateFormData}
          value={value}
          multiple
          required={false}
          labelField="label"
          searchFn={(collection, schema, sortBy) => async (filters) => demoResults}
          allowEmptySearch
        />
      </CompositionWrapDefault>
      {value && value.map((v) => v.label)}
    </div>
  );
};


export const MultipleAltLabel = () => {
  const [value, setValue] = useState(null);
  return (
    <CompositionWrapDefault height="4rem" width="4rem">
      <FieldObjectRef
        uid="demoid"
        unit="demounit"
        updateFormData={alert}
        value={value}
        multiple
        required={false}
        labelField="name"
        searchFn={(collection, schema, sortBy) => async (filters) => demoResults}
        allowEmptySearch
      />
    </CompositionWrapDefault>
  );
};
