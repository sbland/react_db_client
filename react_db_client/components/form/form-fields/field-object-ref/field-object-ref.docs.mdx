---
description: 'A React Form Field Component searching and selecting other collection documents.'
labels: ['form-field', 'field-object-ref']
---

import { FieldObjectRef } from './field-object-ref';
