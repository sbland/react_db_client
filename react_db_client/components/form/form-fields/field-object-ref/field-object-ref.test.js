import '@samnbuk/react_db_client.testing.enzyme-setup';
import React from 'react';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { FilterObjectClass } from '@react_db_client/constants.client-types';
import { SearchAndSelectDropdown } from '@samnbuk/react_db_client.components.search-and-select-dropdown';

import { FieldObjectRef } from './field-object-ref';
import * as compositions from './field-object-ref.composition';

const updateFormData = jest.fn();
const asyncGetDocuments = jest.fn();
const handleSelect = jest.fn();

const demoValue = {
  _id: 'demoitem',
  uid: 'demoitem',
  name: 'demoitem',
  label: 'demoitem',
};

const defaultProps = {
  uid: 'demoid',
  unit: 'demounit',
  updateFormData,
  handleSelect,
  asyncGetDocuments,
  value: demoValue,
  multiple: false,
  required: false,
  labelField: 'name',
  collection: 'collection',
};

describe('FieldObjectRef', () => {
  /* SETUP */
  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });
  beforeEach(() => {
    handleSelect.mockClear();
    updateFormData.mockClear();
    asyncGetDocuments.mockClear();
  });

  /* Tests */
  test('Renders', () => {
    shallow(<FieldObjectRef {...defaultProps} />);
  });
  test('Matches Snapshot', () => {
    const component = shallow(<FieldObjectRef {...defaultProps} />);
    const tree = component.debug();
    expect(tree).toMatchSnapshot();
  });

  describe('Compositions', () => {
    Object.entries(compositions).forEach(([name, Composition]) => {
      test(name, () => {
        mount(<Composition />);
      });
    });
  });

  describe('Unit tests', () => {
    let component;
    beforeEach(() => {
      component = mount(<FieldObjectRef {...defaultProps} />);
    });
    describe('shows a search select field', () => {
      test('should show search and select component', () => {
        expect(component.find(SearchAndSelectDropdown)).toBeTruthy();
      });
    });
    describe('handles select', () => {
      test('should call update fn when search component returns selection', () => {
        const searchComponent = component.find(SearchAndSelectDropdown);
        const selectedId = 'demoid';
        const selectedData = {};
        searchComponent.props().handleSelect(selectedId, selectedData);
        expect(updateFormData).toHaveBeenCalledWith(defaultProps.uid, selectedData);
      });
    });
    describe('Passes props to search and select', () => {
      test('should provide default search function if searchFn override is null', () => {
        const searchAndSelectDropdown = component.find(SearchAndSelectDropdown);
        expect(searchAndSelectDropdown.props()).toMatchSnapshot();
      });
    });
    describe('Searching', () => {
      test('Entering search input should call async get documents', async () => {
        // TODO: Complete this
        const searchField = component.find('input').find('[type="text"]').first();
        act(() => {
          searchField.simulate('focus');
          component.update();
        });
        const searchValue = 'Search Val';
        await act(async () => {
          searchField.simulate('change', { target: { value: searchValue } });
          component.update();
          jest.runOnlyPendingTimers();
          component.update();
          jest.runOnlyPendingTimers();
        });
        const collection = defaultProps.collection;
        const filters = [
          new FilterObjectClass({
            uid: 'search',
            field: defaultProps.labelField,
            value: searchValue,
          }),
        ];
        const schema = '_id';
        const sortBy = defaultProps.labelField;
        expect(asyncGetDocuments).toHaveBeenCalledWith(collection, filters, schema, sortBy);
      });
    });
  });
});
