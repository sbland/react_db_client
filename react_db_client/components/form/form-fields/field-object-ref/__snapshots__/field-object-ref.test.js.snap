// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldObjectRef Matches Snapshot 1`] = `
"<Fragment>
  <SearchAndSelectDropdown searchFunction={[Function (anonymous)]} handleSelect={[Function (anonymous)]} selectionOverride={{...}} allowMultiple={false} returnFieldOnSelect=\\"_id\\" searchFieldTargetField=\\"name\\" labelField=\\"name\\" className=\\"formFieldInput\\" searchFieldPlaceholder=\\"demoitem\\" required={false} allowEmptySearch={false} initialValue=\\"\\" searchDelay={500} valid={true} />
  <span>
    demounit
  </span>
</Fragment>"
`;

exports[`FieldObjectRef Unit tests Passes props to search and select should provide default search function if searchFn override is null 1`] = `
Object {
  "allowEmptySearch": false,
  "allowMultiple": false,
  "className": "formFieldInput",
  "handleSelect": [Function],
  "initialValue": "",
  "labelField": "name",
  "required": false,
  "returnFieldOnSelect": "_id",
  "searchDelay": 500,
  "searchFieldPlaceholder": "demoitem",
  "searchFieldTargetField": "name",
  "searchFunction": [Function],
  "selectionOverride": Array [
    Object {
      "_id": "demoitem",
      "label": "demoitem",
      "name": "demoitem",
      "uid": "demoitem",
    },
  ],
  "valid": true,
}
`;
